### DISCUSSION TITLE
Register changes when an array is mutated

### DISCUSSION THREAD

**Discussion by josegl** [Automattic/mongoose#12171] [Ideas]
Hi guys.
I would like to be able to detect the modifications that happens when an array is updated using the array methods like `splice`, `pop` and so on. This methods mutates the actual array and I'm struggling to be able to detect this changes in the interception of the `markModify` function through a proxy, because as this functions mutate the array, the current value and the "new" value are the same, so any delta can be detected.

Do you have any idea on how this could be achieved?

**Comment by vkarpov15** [comment]
This is a bit tricky. Do you need the state of the array before any changes are made?

**Reply by josegl** [reply]
It would be nice if somehow it would work as the `set` or `$set` functions where the arguments are the `path` being changed and the new value for the given path before applying the new value. However I understand that in this case where the array is mutated, it is very difficult to achieve.

I did a fork  where in the implementation of the `splice` function in `/lib/types/array/methods/index.js` I included a call to `this._markModify()` just the line before this one: `ret = [].splice.apply(arr, vals);` I hoped that doing this the call to `markModified` would bubble up through the model tree, and eventually It would reach my proxy where I intercept the call and I can mark the given path to check for differences. And this is what happens. However, the problem is that when I check the current and the "new" value the array has not updated yet, so there are not any differences to register. If I mark the path to check later, taking the current value as `old_value`, and then I wait to the changes to actually apply, then when I compare the `old_value` with the current value for the given path, the values are the same, because the `old_value` is a pointer to the array that has been actually changed, so they are the same array, and any change can be register.

However as I write this lines I have had an idea: maybe I could try to create a copy instead of a simple assignation for the given path for the `old_value`, and mark it to check it later, so in this case when the array is actually modified, then I could compare different objects.
